## 01. Importing Librabries

In [2]:
import pandas as pd
import numpy as np
import os 

## 02. Importing Data

In [3]:
# Defining path
path = r'/Users/Ayaz/Desktop/Instacart Basket Analysis'

In [4]:
#Importing 'orders_products_merged_updated.pkl' Data set.
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
# Creating subset of first one million rows.
df = df_ords_prods_merge[:1000000]

In [6]:
# checking shape of new subset
df.shape

(1000000, 20)

In [7]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                   15.0         False         196                  1   
2                   21.0         False         196                  1   
3                   29.0         False         196                  1   
4                   28.0         False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days besiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly days        Average orders   
1  Mid-range product  Regularly busy    Slowest days        Average orders   
2  Mid-range product  Regularly busy    Slowest days           Most orders   
3  Mid-range product      Least busy    Slowest days        Average orders   
4  Mid-range product      Least busy    Slowest days           Most orders   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

## 03. Grouping Data with Pandas

In [8]:
df.groupby('product_name')

Always use the groupby() function as part of a series of steps, namely, the following:

1.Split the data into groups based on some criteria.
2.Apply a function to each group separately.
3.Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.
So far, we have only completed the first step—splitting the data into groups (with the group being the “product_name” column). Now, let’s take a look into the second step, which will involve some aggregation!

## 04. Aggregating Data with agg()

### Performing a Single Aggregation.

In [9]:
# Split the data into groups based on “department_id.”
# Apply the agg() function to each group to obtain the mean values for the “order_number” column.

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [10]:
# Different way to find mean of "order_number" 
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

The key difference in syntax between the two methods: when using agg(), put the column you want to aggregate inside the parentheses of the agg() function as an argument. When using mean() (or any other standard aggregation function), simply index the column with square brackets, then follow it with the function you want to use after the dot.

### Performing Mutiple Aggregations.

In [11]:
# Aplying multiple Aggregation to "order_number" column.
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})


order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Task 4.8


### Step 1 - Import Libraries & "ords_prods_merge" Dataframe

In [12]:
# Done (Output [1],[2])

## Step 2 -
### Finding 'mean' of the "order_number" column grouped by "department_id" for entire dataframe.

In [13]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3 - Analyzing the result.

In [14]:
# The Subset 'mean' of the 'order_number' grouped by 'department_id'
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

With the exception of department_id 17 (households), the average orders in the 'order_number' column Subset dataframe are higher than the average across the board. While the minimum average order for the Entire dataframe is 15.215751 for alcohol, the minimum average order for the Subset is 11.294069 for households. The maximum average order for the Subset is 19.463012 for dairy eggs, while the maximum order average for the entire dataframe is 20.197148 for bulk. In the entire dataframe, the average number of "missing" department (21) is 22.902379, which is the highest number.

## Step 4 - Aggregating Data with Transform() & loc()

### Creating a loyalty flag for existing customers.

In [15]:
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [16]:
df_ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   
5                    19.0         False         196                  1   
6                    20.0         False         196                  1   
7                    14.0         False         196                  2   
8                     0.0         False         196                  4   
9                    30.0         False         196                  1   
10                    7.0         False         196                  2   
11                    8.0         False         196                  6   
12                    7.0         False         196                  1   
13                    7.0         False         196                  1   
14                   14.0         False         196                  1   

    reordered  ... aisle_id  department_id  prices  _merge    price_range_loc  \
0           0  ...       77              7     9.0    both  Mid-range product   
1           1  ...       77              7     9.0    both  Mid-range product   
2           1  ...       77              7     9.0    both  Mid-range product   
3           1  ...       77              7     9.0    both  Mid-range product   
4           1  ...       77              7     9.0    both  Mid-range product   
5           1  ...       77              7     9.0    both  Mid-range product   
6           1  ...       77              7     9.0    both  Mid-range product   
7           1  ...       77              7     9.0    both  Mid-range product   
8           1  ...       77              7     9.0    both  Mid-range product   
9           1  ...       77              7     9.0    both  Mid-range product   
10          0  ...       77              7     9.0    both  Mid-range product   
11          1  ...       77              7     9.0    both  Mid-range product   
12          1  ...       77              7     9.0    both  Mid-range product   
13          1  ...       77              7     9.0    both  Mid-range product   
14          1  ...       77              7     9.0    both  Mid-range product   

       busiest_day    busiest_days besiest_period_of_day  \
0   Regularly busy  Regularly days        Average orders   
1   Regularly busy    Slowest days        Average orders   
2   Regularly busy    Slowest days           Most orders   
3       Least busy    Slowest day

## Deriving Column with loc()

In [17]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [18]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [19]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [20]:
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

To check the result instead of .head() functioin we can use a similar syntax to access multiple columns at the same time. For instance, if we wanted to write a head() function that returned the first 30 rows of only the “user_id,” “loyalty_flag,” and “order_number” columns, we could format it as follows:

In [21]:
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(30)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10

## Step 5 - The Basic Statistics for Products Prices Grouped by loyalty_flag.

### Grouping Data with Pandas

In [22]:
# Group by products prices.
df_ords_prods_merge.groupby('prices')

In [23]:
# Split the data into groups based on “loyalty_flag.”
# Apply the agg() function to each group to obtain the mean values for the “price” column.

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal customer    10.386336
New customer      13.294670
Regular customer  12.495717

## Step 6 - Creating Spending Flag on Existing Customers.

In [24]:
# creating "average_price" column using transform() function.
df_ords_prods_merge['average_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [25]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                   15.0         False         196                  1   
2                   21.0         False         196                  1   
3                   29.0         False         196                  1   
4                   28.0         False         196                  1   

   reordered  ... prices  _merge    price_range_loc     busiest_day  \
0          0  ...    9.0    both  Mid-range product  Regularly busy   
1          1  ...    9.0    both  Mid-range product  Regularly busy   
2          1  ...    9.0    both  Mid-range product  Regularly busy   
3          1  ...    9.0    both  Mid-range product      Least busy   
4          1  ...    9.0    both  Mid-range product      Least busy   

     busiest_days besiest_period_of_day busiest_period_of_day max_order  \
0  Regularly days        Average orders        Average orders        10   
1    Slowest days        Average orders        Average orders        10   
2    Slowest days           Most orders           Most orders        10   
3    Slowest days        Average orders        Average orders        10   
4    Slowest days           Most orders           Most orders        10   

   loyalty_flag average_price  
0  New customer      6.367797  
1  New customer      6.367797  
2  New customer      6.367797  
3  New customer      6.367797  
4  New customer      6.367797  

[5 rows x 23 columns]

### Deriving Column with loc() Function.

creating a flag that assigns a "spending" label to a user ID based on its mean product prices.

Criteria: If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [26]:
df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [27]:
df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [29]:
# checkig frequency of spending flag
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

## Step 7 - Creating a Order Frequency Flag

#### The regularity of a user’s ordering behavior according to the median in the “days_since_last_order” column.

In [30]:
# Aggregating data with transform() function
# creating "median_days" column.
df_ords_prods_merge['median_days'] = df_ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [31]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  new_customer  product_id  add_to_cart_order  \
0                    NaN          True         196                  1   
1                   15.0         False         196                  1   
2                   21.0         False         196                  1   
3                   29.0         False         196                  1   
4                   28.0         False         196                  1   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly days   
1          1  ...  Mid-range product  Regularly busy    Slowest days   
2          1  ...  Mid-range product  Regularly busy    Slowest days   
3          1  ...  Mid-range product      Least busy    Slowest days   
4          1  ...  Mid-range product      Least busy    Slowest days   

   besiest_period_of_day busiest_period_of_day max_order  loyalty_flag  \
0         Average orders        Average orders        10  New customer   
1         Average orders        Average orders        10  New customer   
2            Most orders           Most orders        10  New customer   
3         Average orders        Average orders        10  New customer   
4            Most orders           Most orders        10  New customer   

  average_price spending_flag median_days  
0      6.367797   Low spender        20.5  
1      6.367797   Low spender        20.5  
2      6.367797   Low spender        20.5  
3      6.367797   Low spender        20.5  
4      6.367797   Low spender        20.5  

[5 rows x 25 columns]

Creating a flag that assign "frequency" label to user ID based on the median of the “days_since_prior_order” column.

Criteria: If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [32]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [33]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] > 10 & (df_ords_prods_merge['median_days']<=20), 'frequency_flag'] = 'Regular customer'

In [34]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [35]:
# checking the frequency flag
df_ords_prods_merge[['user_id', 'days_since_last_order', 'frequency_flag']].head(30)

user_id  days_since_last_order     frequency_flag
0         1                    NaN   Regular customer
1         1                   15.0   Regular customer
2         1                   21.0   Regular customer
3         1                   29.0   Regular customer
4         1                   28.0   Regular customer
5         1                   19.0   Regular customer
6         1                   20.0   Regular customer
7         1                   14.0   Regular customer
8         1                    0.0   Regular customer
9         1                   30.0   Regular customer
10       15                    7.0  Frequent customer
11       15                    8.0  Frequent customer
12       15                    7.0  Frequent customer
13       15                    7.0  Frequent customer
14       15                   14.0  Frequent customer
15       19                    6.0  Frequent customer
16       19                    8.0  Frequent customer
17       19                    6.0  Frequent customer
18       21                    6.0  Frequent customer
19       31                    7.0  Frequent customer
20       31                    8.0  Frequent customer
21       43                    4.0  Frequent customer
22       43                    3.0  Frequent customer
23       52                    NaN  Frequent customer
24       52                    7.0  Frequent customer
25       52                    7.0  Frequent customer
26       52                    8.0  Frequent customer
27       52                    8.0  Frequent customer
28       52                    7.0  Frequent customer
29       52                    8.0  Frequent customer

In [36]:
# checking the frequency of new flags columns.
df_ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer    21559853
Regular customer     10845001
NaN                         5
Name: frequency_flag, dtype: int64

In [37]:
# checking the shape of dataframe
df_ords_prods_merge.shape

(32404859, 26)

## Step 8 & 9 - Checked and Exporting Data.

In [38]:
# Exporting dataframe as pkl file.
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [39]:
# Check the prices column for further investigation.
df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [41]:
# Checking prices column descriptive statistics.
df['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [43]:
df['prices'].mean()

11.980225638383454

In [44]:
df['prices'].median()

7.4

In [45]:
df['prices'].max()

99999.0

The mean and median are between 7 and 12 (relatively realistic), the max value yields a strange result. How could an item cost $99,999.00? This must be the result of data corruption or some kind of special value within the data.

Solution: Let’s first determine whether there are any values above a certain threshold in data, for instance, anything above the “normal” price of a supermarket item. If we imagined the maximum price anyone would pay in a supermarket were around $100, then anything above that amount could be considered an outlier.

In [48]:
df.loc[df['prices'] > 100]

order_id  user_id  order_number  orders_day_of_week  \
10030345    912404       17            12                   2   
10030346    603376       17            22                   6   
10030347   3264360      135             2                   2   
10030348    892534      135             3                   0   
10030349    229704      342             8                   1   
...            ...      ...           ...                 ...   
29166209   2249946   204099            29                   0   
29166210   2363282   204099            31                   0   
29166211   3181945   204395            13                   3   
29166212   2486215   205227             7                   3   
29166213   2232716   205227            10                   0   

          order_hour_of_day  days_since_last_order  new_customer  product_id  \
10030345                 14                    5.0         False       21553   
10030346                 16                    4.0         False       21553   
10030347                 21                   13.0         False       21553   
10030348                  8                   12.0         False       21553   
10030349                 19                   30.0         False       21553   
...                     ...                    ...           ...         ...   
29166209                  8                    4.0         False       33664   
29166210                  9                    2.0         False       33664   
29166211                 15                    8.0         False       33664   
29166212                 20                    4.0         False       33664   
29166213                 11                    4.0         False       33664   

          add_to_cart_order  reordered  ...     busiest_day    busiest_days  \
10030345                  5          0  ...  Regularly busy  Regularly days   
10030346                  3          1  ...  Regularly busy  Regularly days   
10030347                  6          0  ...  Regularly busy  Regularly days   
10030348                  3          1  ...     Busiest day    Busiest days   
10030349                  9          0  ...  Regularly busy    Busiest days   
...                     ...        ...  ...             ...             ...   
29166209                  1          0  ...     Busiest day    Busiest days   
29166210                  1          1  ...     Busiest day    Busiest days   
29166211                 25          0  ...  Regularly busy    Slowest days   
29166212                  8          0  ...  Regularly busy    Slowest days   
29166213                 23          1  ...     Busiest day    Busiest days   

          besiest_period_of_day  busiest_period_of_day max_order  \
10030345            Most orders            Most orders        40   
10030346            Most orders            Most orders        40   
10030347         Average orders         Average orders         4   
10030348         Average orders         Average orders         4   
10030349         Average orders         Average orders        16   
...                         ...                    ...       ...   
29166209         Average orders         Average orders        39   
29166210            Most orders            Most orders        39   
29166211            Most orders            Most orders        15   
29166212         Average orders         Average orders        12   
29166213            Most orders            Most orders        12   

              loyalty_flag average_price spending_flag median_days  \
10030345  Regular customer    108.648299  High spender         5.0   
10030346  Regular customer    108.648299  High spender         5.0   
10030347      New customer   1154.792308  High spender        12.0   
10030348      New customer   1154.792308  High spender        12.0   
10030349  Regular customer    114.426619  High spender        23.0   
...                    ...           ...           ...         ...   
29166209  Re

From the output, you can see that there are 5,127 rows with outlier observations (prices greater than $100).
For now, let’s mark them as missing since they don’t make sense in terms of the other values in the column. To turn them into NaNs, use the following code:

In [49]:
df.loc[df['prices'] >100, 'prices'] = np.nan

In [50]:
#checking the results
df['prices'].max()

25.0

Now that we’ve replaced all the outliers with missing values, the max value is a much more realistic price point of 25.

Logically, we know that a $25 grocery item is much more likely than a $14,900 item (and especially a $99,999 item!).

In [51]:
# Exporting dataframe as pkl file.
df.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))